File: aoc/aoc23/python/day4.py
def read_lines_from_file(filepath):
    with open(filepath, "r") as f:
        return [line.strip() for line in f.readlines()]


def parse_line(line):
    [card_str, numbers] = line.split(':')
    [winners_str, card_nos_str] = numbers.split('|')

    winners = [int(n) for n in winners_str.split()]
    numbers = [int(n) for n in card_nos_str.split()]

    return {'winners': winners, 'numbers': numbers, 'instances': 1}


def find_winners(card):
    return [num for num in card['numbers'] if num in card['winners']]


def winval(winners):
    num = len(winners)
    if num > 0:
        return 2 ** (num - 1)
    else:
        return 0


def part1(lines):
    cards = [parse_line(line) for line in lines]
    my_winners = [find_winners(card) for card in cards]
    return sum([winval(winners) for winners in my_winners])


ex = read_lines_from_file('./inputs/day4_example.txt')
print(part1(ex))

day4 = read_lines_from_file('./inputs/day4.txt')
print(part1(day4))


def part2(lines):
    cards = [parse_line(line) for line in lines]
    num_cards = len(cards)
    for card_id in range(0, num_cards):
        num_winners = len(find_winners(cards[card_id]))
        instances = cards[card_id]['instances']
        for won_card_id in range(card_id+1, card_id+1+num_winners):
            cards[won_card_id]['instances'] = cards[won_card_id]['instances'] + instances

    return sum([card['instances'] for card in cards])


print(part2(ex))
print(part2(day4))

File: aoc/aoc24/js/day4.js
const fs = require('fs');

function readFromFile(filename) {
    const fileData = fs.readFileSync(filename, 'utf8');
    const lines = fileData.split(/\n/);
    return lines.map(line => line.split(''));
}

const dirs = [
    [1, 0], [1, 1], [0, 1], [-1, 1], [-1, 0], [-1, -1], [0, -1], [1, -1]
]

function find(array, row, col, dir, str){
    let x = col;
    let y = row;
    for (let i=0; i<str.length; ++i){
        x += dir[0];
        y += dir[1];
        if (x < 0 || y<0 || x >= array[0].length || y >=array.length){
            return 0;
        }
        if (array[y][x] !== str[i]){
            return 0;
        }
    }
    return 1;
}

function part1(input) {
    let sum = 0;
    for (let row = 0; row < input.length; ++row) {
        for (let col = 0; col < input[0].length; ++col) {
            if (input[row][col] === 'X') {
                for (let dir of dirs){
                    sum += find(input, row, col, dir, 'MAS');
                }
            }
        }
    }
    return sum;
}

function part2(input) {
    let sum = 0;
    for (let row = 1; row < input.length - 1; ++row) {
        for (let col = 1; col < input[0].length - 1; ++col) {
            if (input[row][col] === 'A') {
                const tl = input[row-1][col-1];
                const br = input[row+1][col+1];
                if ((tl === 'M' && br === 'S') || (tl === 'S' && br === 'M')){
                    const tr = input[row-1][col+1];
                    const bl = input[row+1][col-1];
                    if ((tr === 'M' && bl === 'S') || (tr === 'S' && bl === 'M')){
                        sum++;
                    }
                }
            }
        }
    }
    return sum;
}

const input = readFromFile('./inputs/day4.txt');

console.log(part1(input));
console.log(part2(input));
